packages/core/src/migration/index.ts

```diff
@@ -0,0 +1,134 @@
+import { readdir, readFile } from 'fs/promises';
+import path from 'path';
+
+import { LogtoConfig } from '@logto/schemas';
+import { conditionalString } from '@silverhand/essentials';
+import chalk from 'chalk';
+import { DatabasePool, sql } from 'slonik';
+import { raw } from 'slonik-sql-tag-raw';
+
+import {
+  databaseVersionKey,
+  logtoConfigsTable,
+  logtoConfigsTableFilePath,
+  migrationFilesDirectory,
+} from './constants';
+import { DatabaseVersion, databaseVersionGuard, MigrationScript } from './types';
+import { compareVersion, getVersionFromFileName, migrationFileNameRegex } from './utils';
+
+const isLogtoConfigsTableExists = async (pool: DatabasePool) => {
+  const [result] = await pool.anyFirst<[boolean]>(sql`
+    select exists (
+      select from 
+        pg_tables
+      where 
+        tablename = ${logtoConfigsTable}
+      );
+  `);
+
+  return result;
+};
+
+const getCurrentDatabaseVersion = async (pool: DatabasePool) => {
+  try {
+    const query = await pool.maybeOne<LogtoConfig>(
+      sql`select * from ${sql.identifier([logtoConfigsTable])} where key=${databaseVersionKey}`
+    );
+    const databaseVersion = databaseVersionGuard.parse(query?.value);
+
+    return databaseVersion.version;
+  } catch {
+    return null;
+  }
+};
+
+const createLogtoConfigsTable = async (pool: DatabasePool) => {
+  const tableQuery = await readFile(logtoConfigsTableFilePath, 'utf8');
+  await pool.query(sql`${raw(tableQuery)}`);
+};
+
+const updateDatabaseVersion = async (pool: DatabasePool, version: string) => {
+  if (!(await isLogtoConfigsTableExists(pool))) {
+    await createLogtoConfigsTable(pool);
+  }
+
+  const value: DatabaseVersion = {
+    version,
+    updatedAt: new Date().toISOString(),
+  };
+
+  await pool.query(
+    sql`
+      insert into ${sql.identifier([logtoConfigsTable])} (key, value) 
+        values (${databaseVersionKey}, ${JSON.stringify(value)})
+        on conflict (key) do update set value=excluded.value
+    `
+  );
+};
+
+const getMigrationFiles = async () => {
+  const directory = await readdir(migrationFilesDirectory);
+  const files = directory.filter((file) => migrationFileNameRegex.test(file));
+
+  return files;
+};
+
+export const getUndeployedMigrations = async (pool: DatabasePool) => {
+  const databaseVersion = await getCurrentDatabaseVersion(pool);
+  const files = await getMigrationFiles();
+
+  if (!databaseVersion) {
+    return files;
+  }
+
+  return files
+    .filter((file) => compareVersion(getVersionFromFileName(file), databaseVersion) > 0)
```

Comment:
if we directly return `files` will be unsorted
```suggestion
  return files
    .filter((file) => !databaseVersion || compareVersion(getVersionFromFileName(file), databaseVersion) > 0)
```